crates/nomination/src/tests.rs

```diff
@@ -32,3 +34,65 @@ fn should_deposit_against_valid_vault() {
         assert_ok!(Nomination::_deposit_collateral(&ALICE, &BOB.account_id, 100));
     })
 }
+
+#[test]
+fn should_not_withdraw_collateral() {
+    use orml_traits::MultiCurrency;
+
+    run_test(|| {
+        assert_ok!(Tokens::deposit(
+            ALICE.currencies.collateral,
+            &ALICE.account_id,
+            24_609_778_406_619_232
+        ));
+        VaultRegistry::_set_system_collateral_ceiling(ALICE.currencies, u128::MAX);
+        assert_ok!(VaultRegistry::register_public_key(
+            RuntimeOrigin::signed(ALICE.account_id),
+            vault_registry::BtcPublicKey::dummy()
+        ));
+        assert_ok!(VaultRegistry::register_vault(
+            RuntimeOrigin::signed(ALICE.account_id),
+            ALICE.currencies,
+            24_609_778_406_619_232
+        ));
+
+        staking::SlashPerToken::<Test>::insert(0, ALICE, FixedI128::from_inner(25_210_223_519_649_666));
+        staking::SlashTally::<Test>::insert(
+            0,
+            (ALICE, ALICE.account_id),
+            FixedI128::from_inner(100_834_580_684_768_029_667_333_677_168),
+        );
+        staking::Stake::<Test>::insert(
+            0,
+            (ALICE, ALICE.account_id),
+            FixedI128::from_inner(3_999_749_570_096_999_994_120_799_432_121),
+        );
+        staking::TotalCurrentStake::<Test>::insert(
+            0,
+            ALICE,
+            FixedI128::from_inner(3_999_749_570_097_000_000_000_000_000_000),
+        );
+        staking::TotalStake::<Test>::insert(
+            0,
+            ALICE,
+            FixedI128::from_inner(3_999_749_570_096_999_994_120_799_432_121),
+        );
+
+        // should not withdraw all
+        assert_err!(
+            Nomination::_withdraw_collateral(&ALICE, &ALICE.account_id, Some(3999749570097), 0),
+            staking::Error::<Test>::InsufficientFunds
+        );
+
+        // should withdraw all
+        assert_ok!(Nomination::_withdraw_collateral(&ALICE, &ALICE.account_id, None, 0));
+
+        // stake is now zero
+        assert_ok!(ext::staking::compute_stake::<Test>(&ALICE, &ALICE.account_id), 0);
+        // TODO: TotalCurrentStake is non-zero due to rounding
+        assert_ok!(
+            VaultRegistry::get_backing_collateral(&ALICE),
+            Amount::new(1, ALICE.collateral_currency())
+        );
```

Comment:
We can either change the staking API to clear that field internally or I would also be in favor of leaving this for when we remove the `staking` pallet.

Reply:
Hmm TotalCurrentStake is used to determine vault collateral, as well as their stakes in reward pools. What are the vault's `reserved` tokens at this point? I'm a little bit afraid that accounting mismatches can come back to bite us

Reply:
Resolved by implementing `withdraw_all_stake` but I also opted to cleanup the `TotalCurrentStake` storage since there is still a small discrepancy (`0.000005879200567879`).